src/Stratis.Bitcoin/P2P/PeerSelector.cs

```diff
@@ -195,11 +198,24 @@ private IEnumerable<PeerAddress> SelectPreferredPeers()
                 return attempted;
             }
 
-            // If all the selection criteria failed to return a set of peers,
-            // then let the caller try again.
-            else
-                this.logger.LogTrace("[RETURN_NO_PEERS]");
+            List<PeerAddress> notBanned = this.NotBanned().ToList();
```

Comment:
@noescape00 this is much better yes but I would really prefer this in a method so that we can unit test it as it is quite an important bit of code ;)

You can split this logic to a property or method called something like `HasAllPeersReachedConnectionThreshold`:

```
int attemptedReachedThresholdCount = notBanned.Count(p => p.ConnectionAttempts == PeerAddress.AttemptThreshold);
bool areAllPeersReachedThreshold = attemptedReachedThresholdCount == notBanned.Count;
```

and then, if so, create a method called `ResetConnectionAttemptsOnAllPeers()` which does:

```
this.logger.LogTrace("Resetting attempts for {0} addresses.", attemptedReachedThresholdCount);

// Reset attempts for all the peers since we've ran out of options.
foreach (PeerAddress peer in notBanned)
     peer.ResetAttempts();

this.logger.LogTrace("(-)[RESET_ATTEMPTS]");
return notBanned;
```

So after it will look like this:

```
if (HasAllPeersReachedConnectionThreshold)
    ResetConnectionAttemptsOnAllPeers();
```

Reply:
This way its very easy to unit test.

Reply:
Ok, makes sense. 

Fixed, added unit tests.